src/Chessboard.cpp

```diff
@@ -0,0 +1,313 @@
+            }
+        }
+    }
+    for (int i = kingRow - 1; i >= 0; i--) {
+        Piece* piece = board_[i][kingCol];
+        if (piece != nullptr) {
+            if (piece->getColor() == color) {
+                break;
+            }
+            else if (piece->getType() == PieceType::Rook || piece->getType() == PieceType::Queen) {
+                return true;
+            }
+            else {
+                break;
+            }
+        }
+    }
+
+    // check for bishops and queens attacking diagonally
+    int i = kingRow + 1, j = kingCol + 1;
+    while (i < 8 && j < 8) {
+        Piece* piece = board_[i][j];
+        if (piece != nullptr) {
+            if (piece->getColor() == color) {
+                break;
+            }
+            else if (piece->getType() == PieceType::Bishop || piece->getType() == PieceType::Queen) {
+                return true;
+            }
+            else {
+                break;
+            }
+        }
+        i++;
+        j++;
+    }
+    i = kingRow - 1, j = kingCol - 1;
+    while (i >= 0 && j >= 0) {
+        Piece* piece = board_[i][j];
+        if (piece != nullptr) {
+            if (piece->getColor() == color) {
+                break;
+            }
+            else if (piece->getType() == PieceType::Bishop || piece->getType() == PieceType::Queen) {
+                return true;
+            }
+            else {
+                break;
+            }
+        }
+        i--;
+        j--;
+    }
+    i = kingRow + 1, j = kingCol - 1;
+    while (i < 8 && j >= 0) {
+        Piece* piece = board_[i][j];
+        if (piece != nullptr) {
+            if (piece->getColor() == color) {
+                break;
+            }
+            else if (piece->getType() == PieceType::Bishop || piece->getType() == PieceType::Queen) {
+                return true;
+            }
+            else {
+                break;
+            }
+        }
+        i++;
+        j--;
+    }
+    i = kingRow - 1, j = kingCol + 1;
+    while (i >= 0 && j < 8) {
+        Piece* piece = board_[i][j];
+        if (piece != nullptr) {
+            if (piece->getColor() == color) {
+                break;
+            }
+            else if (piece->getType() == PieceType::Bishop || piece->getType() == PieceType::Queen) {
+                return true;
+            }
+            else {
+                break;
+            }
+        }
+        i--;
+        j++;
+    }
+
+    // check for knights attacking
+    int knightMoves[8][2] = { {-2, 1}, {-1, 2}, {1, 2}, {2, 1}, {2, -1}, {1, -2}, {-1, -2}, {-2, -1} };
+    for (int k = 0; k < 8; k++) {
+        int row = kingRow + knightMoves[k][0];
+        int col = kingCol + knightMoves[k][1];
+        if (row >= 0 && row < 8 && col >= 0 && col < 8) {
+            Piece* piece = board_[row][col];
+            if (piece != nullptr && piece->getType() == PieceType::Knight && piece->getColor() != color) {
+                return true;
+            }

```

Comment:
Instead of using nested loops to check all possible knight moves, you can simplify it by directly checking the eight possible knight moves around the king.